ykcompile/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+#![feature(proc_macro_hygiene)]
+#![feature(test)]
+
+#[macro_use]
+extern crate dynasm;
+extern crate dynasmrt;
+extern crate test;
+
+use std::collections::HashMap;
+use std::mem;
+
+use yktrace::tir::{TirTrace, TirOp, Statement, Rvalue, Operand, Constant, ConstantInt,
+UnsignedInt};
+
+use dynasmrt::DynasmApi;
+
+/// A compiled SIRTrace.
+pub struct CompiledTrace {
+    /// A compiled trace.
+    mc: dynasmrt::ExecutableBuffer
+}
+
+impl CompiledTrace {
+    pub fn execute(&self) -> u64 {
+        // For now a compiled trace always returns whatever has been left in register RAX. We also
+        // assume for now that this will be a u64.
+        let func: fn() -> u64 = unsafe {
+            mem::transmute(self.mc.ptr(dynasmrt::AssemblyOffset(0)))
+        };
+        func()
+    }
+}
+
+/// The TraceCompiler takes a SIRTrace and compiles it to machine code. Returns a CompiledTrace.
+pub struct TraceCompiler {
+    asm: dynasmrt::x64::Assembler,
+    available_regs: Vec<u8>,
+    assigned_regs: HashMap<u32, u8>
```

Comment:
Is that local index to register number?

We should add comments.

Reply:
Done.